### PR TITLE
5767 add poll waiting step

### PR DESCRIPTION
**Pull request by TipzCM**

### CODE REVIEW

File: hapi-fhir-storage-batch2/src/main/java/ca/uhn/fhir/batch2/model/WorkChunkStatusEnum.java [hunk 1]
```diff
@@ -32,11 +32,30 @@
...
+	/**
+	 * The state of workchunks that are doing work.
+	 */
 	IN_PROGRESS,
+	/**
+	 * A workchunk status for workchunks that are doing long-polling work
+	 * that will not complete in a reasonably short amount of time
+	 */
+	POLL_WAITING,
 	ERRORED,
```
[the remark this answers is not in the document]
Comment (michaelabuckley) [reply hapi-fhir-storage-batch2/src/main/java/ca/uhn/fhir/batch2/model/WorkChunkStatusEnum.java:54]: Please add to the next-states bit down at the bottom.

**Review by michaelabuckley — APPROVED** [review]
looks ok.  some minor quibbles.  One big gap is the missing spec-test for the state transitions.

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/batch2/JpaJobPersistenceImpl.java [hunk 2]
```diff
@@ -488,6 +511,16 @@ public WorkChunk createWorkChunk(WorkChunk theWorkChunk) {
 		return toChunk(myWorkChunkRepository.save(Batch2WorkChunkEntity.fromWorkChunk(theWorkChunk)));
 	}
 
+	@Override
+	public void updateWorkChunkToStatus(
+			String theChunkId, WorkChunkStatusEnum theOldStatus, WorkChunkStatusEnum theNewStatus) {
```
[the remark this answers is not in the document]
Comment (TipzCM) [reply hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/batch2/JpaJobPersistenceImpl.java:516]: removed

File: hapi-fhir-storage-batch2/src/main/java/ca/uhn/fhir/batch2/api/RetryChunkLaterException.java [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+	 * For now, it's a constant, but we hold it here in
+	 * case we want to change this behaviour in the future.
+	 */
+	private int myNextPollDelayMs = ONE_MIN;
+
+	public RetryChunkLaterException() {
+		this(ONE_MIN);
+	}
+
+	public RetryChunkLaterException(int theMsDelay) {
```
[the remark this answers is not in the document]
Comment (TipzCM) [reply hapi-fhir-storage-batch2/src/main/java/ca/uhn/fhir/batch2/api/RetryChunkLaterException.java:22]: changed to duration

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/data/IBatch2WorkChunkRepository.java [hunk 4]
```diff
@@ -75,6 +76,23 @@ void updateChunkStatusAndClearDataForEndSuccess(
...
+	@Query(
+			"UPDATE Batch2WorkChunkEntity e SET e.myStatus = :status, e.myNextPollTime = :nextPollTime, e.myPollAttempts = e.myPollAttempts + 1 WHERE e.myId = :id")
+	int updateWorkChunkNextPollTime(
+			@Param("id") String theChunkId,
+			@Param("status") WorkChunkStatusEnum theStatus,
+			@Param("nextPollTime") Date theNextPollTime);
+
+	@Modifying
+	@Query(
+			"UPDATE Batch2WorkChunkEntity e SET e.myStatus = :status, e.myNextPollTime = null WHERE e.myInstanceId = :instanceId AND e.myStatus IN(:states) AND e.myNextPollTime IS NOT NULL AND e.myNextPollTime <= :pollTime")
```
[the remark this answers is not in the document]
Comment (TipzCM) [reply hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/data/IBatch2WorkChunkRepository.java:89]: updated

File: hapi-fhir-storage-batch2/src/main/java/ca/uhn/fhir/batch2/model/WorkChunkStatusEnum.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (TipzCM) [reply hapi-fhir-storage-batch2/src/main/java/ca/uhn/fhir/batch2/model/WorkChunkStatusEnum.java:54]: added

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/batch2/JpaJobPersistenceImpl.java [hunk 5]
```diff
@@ -350,6 +361,18 @@ public WorkChunkStatusEnum onWorkChunkError(WorkChunkErrorEvent theParameters) {
 		});
 	}
 
+	@Override
+	public void onWorkChunkPollDelay(String theChunkId, int thePollDelayMs) {
```
[the remark this answers is not in the document]
Comment (TipzCM) [reply hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/batch2/JpaJobPersistenceImpl.java:365]: added

File: hapi-fhir-storage-batch2/src/main/java/ca/uhn/fhir/batch2/api/IJobPersistence.java [hunk 6]
```diff
@@ -93,6 +93,15 @@ List<JobInstance> fetchInstances(
...
 	void enqueueWorkChunkForProcessing(String theChunkId, Consumer<Integer> theCallback);
 
+	/**
+	 * Updates all Work Chunks in POLL_WAITING if their nextPollTime <= now
+	 * for the given Job Instance.
+	 * @param theInstanceId the instance id
+	 * @return the number of updated chunks
+	 */
+	@Transactional
+	int updatePollWaitingChunksForJobIfReady(String theInstanceId);
```
[the remark this answers is not in the document]
Comment (TipzCM) [reply hapi-fhir-storage-batch2/src/main/java/ca/uhn/fhir/batch2/api/IJobPersistence.java:103]: added

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/data/IBatch2WorkChunkRepository.java [hunk 7]
```diff
@@ -49,7 +49,8 @@ public interface IBatch2WorkChunkRepository
 	@Query("SELECT new Batch2WorkChunkEntity("
 			+ "e.myId, e.mySequence, e.myJobDefinitionId, e.myJobDefinitionVersion, e.myInstanceId, e.myTargetStepId, e.myStatus,"
 			+ "e.myCreateTime, e.myStartTime, e.myUpdateTime, e.myEndTime,"
-			+ "e.myErrorMessage, e.myErrorCount, e.myRecordsProcessed, e.myWarningMessage"
+			+ "e.myErrorMessage, e.myErrorCount, e.myRecordsProcessed, e.myWarningMessage,"
+			+ "e.myNextPollTime, e.myPollAttempts"
```
[the remark this answers is not in the document]
Comment (TipzCM) [reply hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/data/IBatch2WorkChunkRepository.java:53]: I don't think so for this version.

Could be something we add in the future though

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/batch2/JpaJobPersistenceImpl.java [hunk 8]
```diff
@@ -350,6 +361,18 @@ public WorkChunkStatusEnum onWorkChunkError(WorkChunkErrorEvent theParameters) {
...
+		int updated = myWorkChunkRepository.updateWorkChunkNextPollTime(
+				theChunkId,
+				WorkChunkStatusEnum.POLL_WAITING,
```
[the remark this answers is not in the document]
Comment (TipzCM) [reply hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/batch2/JpaJobPersistenceImpl.java:368]: I think only IN_PROGRESS (maybe ERRORED?)

But for now, i've made it only IN_PROGRESS.

If something is POLL_WAITING, we transition it to READY before retrying (so the state machine is always IN_PROGRESS -> POLL_WAITING, POLL_WAITING -> READY

**Pull request closed without merging**